### ISSUE TITLE
Error_computation

### ISSUE THREAD

**Issue by aradermacher** [BAMresearch/PGDrome#5]
general error computation function to check PGD model against reference model (analytic, fem) for a randomly sampled set of parameters

**Comment by aradermacher** [comment]
Can we put the whole loop which is in the moment implemented in the example into the error function too?
Maybe use the already existing function "error_computation" in the PGD model?

**Comment by agarllo** [comment]
I was working on it and I have almost implemented for one of the test (not running jet). But I am thinking that it could be useful leaving the loop in the main script (but in a cleaner way) and give the user the option to select the Exact solution. Now it is being computed through FEM and Analitical eq. But in case the results are previously computed or need to define any other extra parameter, I have doubts if it could be flexible enough. When I finish implementig, I upload the new code and we discus it

**Comment by aradermacher** [comment]
I would suggest something like a reference model class like this:

class Reference():
    def __init__(self,params, other things I need):

        self.mesh = 
        self.V = 
        self.params = 
        .....

   def functions_I_need():
     return ...

   def __call__(self, values):
        compute reference solution for values
        return ref_sol

This class has to be given in each test for each problem. In the pgdmodel.error_computation_samples(..,Reference,..) function the reference solution can then be computed as Reference(current_sampled_values)

**Comment by aradermacher** [comment]
You use the qmc modul for the LatinHypercube sampling. but it is not imported! From scipy?

Can we use the same names for the evaluation function as before e.g. seq_fp should be free_dim ( fixed_dim, free_dim, coord, attri)

Also the fixed_fim is not everytime == 0 as well as attr --> Has to be added in the input of the class

Can we initialize the class PGDErrorComputation with all required inputs, so that it is also possible to just call 
`
error_uPGD = PGDErrorComputation(all inputs like PGDclass, ReferenceClass, evaluationsarguments for PGD, ranges for samples )
error_uPGD.evaluate_error(n_samples)
`

**Comment by aradermacher** [comment]
Why do the reference solution need the iteration parameter?

The mesh is included in the fenics solution function: u.function_space().mesh() -> not necessary to return!

fenics has also a norm computation function: e.g. l2 norm between fencis functions: dolfin.norm(v,'l2')

What's with the old error_computation function -> delete/remove/reuse?

**Comment by agarllo** [comment]
I have changed most of the things you pointed. I am having problems to compute the error using dolfin.norm

**Comment by DoStrobl** [comment]
Can you describe the problem? I assume you have looked into the [documentation](https://fenicsproject.org/olddocs/dolfin/1.5.0/python/programmers-reference/fem/norms/norm.html)?
Depending on what you are doing you can also use the [numpy norm](https://numpy.org/doc/stable/reference/generated/numpy.linalg.norm.html).

**Comment by agarllo** [comment]
Even in the uploaded code is computed using numpy, I computed in the two ways 1) Using dolfin.norm 2) numpy to ensure that is working well. The problem is that one of the test have an analytical solution and it is an NDARRAY variable. I need to project this nodal results in order to get a VECTOR variable and compute the error using dolfin.norm. The problem is more a matter of the type of variables than error computation

**Comment by DoStrobl** [comment]
I don't fully get the problem but maybe this helps:
Use the mesh for the analytical solution to create a function space
`V = fenics.FunctionSpace(mesh, "CG", 2)`
and define a function on it
`u = fenics.Function(V)`.
Then you can manipulate the values of the Function u via
`u.vector()[:] = ...`.
If the ordering doesn't fit check out the mapping
`_map1 = dolfin.vertex_to_dof_map(V)` 
`_map2 = dolfin.dof_to_vertex_map(V)`
This can then be used to get the values in the correct ordering via
`u.vector()[dolfin.vertex_to_dof_map(V)] = ...`.
Finaly the dolfin.norm can be used on the function u representing the analytical solution as a function.

**Comment by aradermacher** [comment]
**First** to the projection: That's possible by 
```
u_ana = dolfin.Function(V)
u_ana.vector()[:]=np.array(analytic_solution)
```
The order of the np.array has to be the same as the dofs!

**Second**: 

- Why do you use the seq_fp variable -> That's a solver parameter and should have nothing to do with the evaluation, or?
- The meshes are already included in the PGDmodel class (pgd.mesh[dim])
- The functions space self.Vs is not used!
- In Line 1432 the self.fixed_dim variable is missing and the attribute should also be given as self.attr!
- The sampling boundary setting is weird. Why to you initialize the boundaries with self.seq_fp and the fill in a loop of len self.free_dim? Should be in both cases self.free_dim! Further, I would like to have the option to give the ranges to (for the free variables, maybe the user is interested in a specific smaller parameter space!) and just if it's not given the ranges is computed by the min/max of the mesh (Only possible for 1D meshes, by the way. I should assert an Error if there are more dimensional meshes at that point)
- We need some more description of the function: e.g. the data_test has do be an array, even it is just one point .../ the pgd_model has to include the fenics_interpolation functions (e.g. if it is read by a file, this has to be done before calling errorcomputation!)
- If you sort the free_dim array in Line 1382 you have to be carefully with the evaluation of pgd - could be broken then. Why has the aux_range be sorted?
- for the error computation decision (function or array) you could look here https://github.com/BAMresearch/fenics-constitutive/blob/module/module/truss_convergence.py   e.g.:
```
         try:
            # numpy ?
            err = np.linalg.norm(u_fem - u_correct)
        except TypeError:
            err = df.errornorm(u_correct, u_fem, norm_type="l2")
```
- old error function has to be deleted!
- tests are not working!!
- in test we should also have the old version where we compute the pgd solution at a given point and compute an error measure plus the new error computation with a given data_test set (may be the same one point), plus for a randomly sampled set (not required for all tests but at least for test_elastic.py and test_Elastic2D_4variables.py)

**Comment by agarllo** [comment]
I think everything is changed and it should be running well

**Comment by aradermacher** [comment]
- delete pgd_path and read_pgd as it's not implemented and not used
- don't use print use the logger as shown in class pgd or class solver and if it's an error do self.logger.error('note') and raise Value Error('note') see e.g. line 956 in model.py
- Good solution for given points!

--> after changing these things please make a pull request and merge the version with the master (I think there were changes in the master during your work)
-->afterwards we will create a new issue and branch to apply this new feature in the other tests (solver_problem/ thermo/ pgdclass_dolfin)
